lib/spack/spack/database.py

```diff
@@ -1648,32 +1658,49 @@ def is_occupied_install_prefix(self, path):
         with self.read_transaction():
             return path in self._installed_prefixes
 
-    @property
-    def unused_specs(self):
-        """Return all the specs that are currently installed but not needed
-        at runtime to satisfy user's requests.
-
-        Specs in the return list are those which are not either:
-            1. Installed on an explicit user request
-            2. Installed as a "run" or "link" dependency (even transitive) of
-               a spec at point 1.
+    def all_hashes(self):
+        """Return dag hash of every spec in the database."""
+        return list(self._data.keys())
+
+    def unused_specs(
+        self,
+        root_hashes: Optional[Container[str]] = None,
+        deptypes: Optional[Tuple[str, ...]] = None,
```

Comment:
```suggestion
        depflag: dt.DepFlag = dt.LINK | dt.RUN,
```

is better in that it doesn't require validation

Reply:
fixed.